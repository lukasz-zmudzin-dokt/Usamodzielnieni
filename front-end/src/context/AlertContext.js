import React, { useState, createContext } from "react";

export const AlertContext = createContext({
  open: false,
  message: "",
  setOpen: () => {},
});

export const AlertProvider = (props) => {
  const [open, setOpen] = useState(false);
  const [message, setMessage] = useState("");
  const [variant, setVariant] = useState("danger") // success
  const data = {
    open,
    message,
    variant,
    changeVisibility: (newOpen) => setOpen(newOpen),
    changeMessage: (newMessage,newVariant = variant) => {
      setOpen(false);
      setMessage(newMessage);
      setVariant(newVariant);
    },
    showAlert:  (newMessage,newVariant) => {
      data.changeMessage(newMessage,newVariant);
      data.changeVisibility(true);
    }
  };
  return <AlertContext.Provider value={data} {...props} />;
};
