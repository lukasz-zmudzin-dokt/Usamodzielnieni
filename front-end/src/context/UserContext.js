import React, { useState } from "react";
import Cookies from "universal-cookie";

export const UserContext = React.createContext({
  type: undefined, // Employer,Standard,Staff
  token: undefined,
  login: () => {},
  logout: () => {}
});
const cookies = new Cookies();

export const UserProvider = props => {
  const [type, setType] = useState(cookies.get("type"));
  const [token, setToken] = useState(cookies.get("token"));
  const user = {
    token,
    type,
    login: (newToken, newType) => {
      cookies.set("token", newToken, { path: "/" });
      cookies.set("type", newType, { path: "/" });
      setToken(newToken);
      setType(newType);
    },
    logout: () => {
      cookies.remove("token", { path: "/" });
      cookies.remove("type", { path: "/" });
      setToken(undefined);
      setType(undefined);
    }
  };
  return <UserContext.Provider value={user} {...props} />;
};
