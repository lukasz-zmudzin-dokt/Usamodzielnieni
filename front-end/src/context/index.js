export { UserContext, UserProvider } from "context/UserContext";
export { AlertContext, AlertProvider } from "context/AlertContext";
export { ChatContext, ChatProvider } from "context/ChatContext";
export {
  NotificationsContext,
  NotificationsProvider,
} from "context/NotificationsContext";
