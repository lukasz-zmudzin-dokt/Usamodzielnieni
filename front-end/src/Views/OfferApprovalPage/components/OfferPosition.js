import React, { useContext, useEffect, useState } from "react";
import { UserContext } from "context";
import { Alert, Button, Card, ListGroup, Row } from "react-bootstrap";
import { DetailsItem } from "components";
import { getOffer, setOfferApproved, setOfferRejected } from "Views/OfferApprovalPage/apiCalls";

const OfferPosition = ({ offer, activeOffer }) => {

    const context = useContext(UserContext);
    const [offerDetails, setOfferDetails] = useState([]);
    const [loading, setLoading] = useState(false);
    const [error, setError] = useState(false);
    const [approved, setApproved] = useState(false);
    const [rejected, setRejected] = useState(false);

    useEffect(() => {
        const loadOfferDetails = async (token, offerId) => {
            /*#TODO połączyć z api jak będzie
            setLoading(true);
            try {
                let res = await getOffer(token, offerId);
                setOfferDetails(res);
                setLoading(false);
            } catch (e) {
                setError(true);
                setLoading(false);
            }*/
        };
        if(activeOffer === offer.id) {
            loadOfferDetails(context.token, offer.id);
        }
    }, [context.token, activeOffer]);

    const approveOffer = async (e, token, offerId) => {
        e.preventDefault();
        /*#TODO połączyć z api jak będzie
        try {
            let res = setOfferApproved(token, offerId);
        } catch (e) {
            setError(true);
        }*/
    };

    const rejectOffer = async (e, token, offerId) => {
        e.preventDefault();
        /*#TODO połączyć z api jak będzie
        try {
            let res = setOfferRejected(token, offerId);
        } catch (e) {
            setError(true);
        }*/
    };

    const message = loading ? (
        <Alert className="mb-0" variant="info">Ładuję...</Alert>
    ) : error ? (
        <Alert className="mb-0" variant="danger">Ups, wystąpił błąd...</Alert>
    ) : approved ? (
        <Alert className="mb-0" variant="success">Oferta zatwierdzona pomyślnie.</Alert>
    ) : rejected ? (
        <Alert className="mb-0" variant="success">Oferta odrzucona pomyślnie.</Alert>
    ) : null;

    return (
        message ? message : (
            <Card.Body>
                i tu będą szczegóły oferty
            </Card.Body>
        )
    );

};

export default OfferPosition;