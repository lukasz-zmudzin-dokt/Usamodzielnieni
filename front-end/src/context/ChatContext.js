import React, {
  createContext,
  useEffect,
  useState,
  useContext,
  useCallback,
} from "react";
import proxy from "config/api";
import { UserContext } from "context";

const getChats = async (token, filters) => {
  let url = `${proxy.chat}?page=${filters.page}&page_size=${filters.pageSize}`;
  const headers = {
    Authorization: "Token " + token,
    "Content-Type": "application/json",
  };

  const response = await fetch(url, { method: "GET", headers });

  if (response.status === 200) {
    return response.json().then((chats) => chats);
  } else {
    throw response.status;
  }
};

export const ChatContext = createContext(null);

export const ChatProvider = (props) => {
  const user = useContext(UserContext);
  const [chats, setChats] = useState([]);
  const [isChatsLoading, setIsChatsLoading] = useState(false);
  const [error, setError] = useState(false);
  const [count, setCount] = useState(0);
  const [filters] = useState({
    page: 1,
    pageSize: 10,
  });
  const [activePage, setActivePage] = useState(1);

  useEffect(() => {
    if (user.token) {
      const loadChats = async (token) => {
        setActivePage(1);
        setError(false);
        setIsChatsLoading(true);
        let loadedChats;
        try {
          loadedChats = await getChats(token, filters);
        } catch (e) {
          console.log(e);
          loadedChats = { count: 0, results: [] };
          setError(true);
        }
        setChats(loadedChats.results);
        setCount(loadedChats.count);
        setIsChatsLoading(false);
      };
      loadChats(user.token);
    }
  }, [filters, user.token]);

  const loadMoreMessages = async () => {
    let res;
    try {
      res = await getChats(user.token, { ...filters, page: activePage + 1 });
      setActivePage(activePage + 1);
    } catch (e) {
      setError(true);
      return;
    }

    setChats([...chats, ...res.results]);
  };

  const loadMessages = async (token) => {
    setActivePage(1);
    let res;
    try {
      res = await getChats(token, filters);
    } catch (e) {
      setError(true);
      return;
    }

    setChats(res.results);
  };

  const data = {
    chats,
    error,
    count,
    loadMoreMessages,
    isChatsLoading,
    loadMessages,
  };

  return <ChatContext.Provider value={data} {...props} />;
};
