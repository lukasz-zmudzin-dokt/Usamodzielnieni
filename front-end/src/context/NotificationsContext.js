import React, {
  useState,
  useContext,
  useEffect,
  useRef,
  createContext,
} from "react";
import { UserContext } from "context";
import proxy from "config/api";
import { paths } from "constants/paths";
import { userTypes } from "constants/userTypes";
import { compile } from "path-to-regexp";

const getNotifications = async (token, type, next) => {
  let url = proxy.notification + "list/all" + (next ? `?page=${next}` : "");
  const headers = {
    Authorization: "Token " + token,
    "Content-Type": "application/json",
  };

  const response = await fetch(url, { method: "GET", headers });

  if (response.status !== 200) {
    throw response.status;
  }

  return response
    .json()
    .then((notifications) => mapNotifications(notifications, type));
};

const getNotificationsCount = async (token) => {
  let url = proxy.notification + "count/unread";
  const headers = {
    Authorization: "Token " + token,
    "Content-Type": "application/json",
  };

  const response = await fetch(url, { method: "GET", headers });

  if (response.status !== 200) {
    throw response.status;
  }

  return (await response.json()).unread_count;
};

const markAsRead = async (token) => {
  let url = proxy.notification + "list/all/mark-as-read/";
  const headers = {
    Authorization: "Token " + token,
    "Content-Type": "application/json",
  };

  const response = await fetch(url, { method: "POST", headers });

  if (response.status !== 200) {
    throw response.status;
  }

  return;
};

const deleteNotification = async (token, id) => {
  let url = proxy.notification + `delete/${id}/`;
  const headers = {
    Authorization: "Token " + token,
    "Content-Type": "application/json",
  };

  const response = await fetch(url, { method: "DELETE", headers });

  if (response.status !== 200) {
    throw response.status;
  }

  return;
};

const deleteNotifications = async (token) => {
  let url = proxy.notification + "list/all/delete/";
  const headers = {
    Authorization: "Token " + token,
    "Content-Type": "application/json",
  };

  const response = await fetch(url, { method: "DELETE", headers });

  if (response.status !== 200) {
    throw response.status;
  }

  return;
};

const mapNotifications = (notifications, type) => ({
  notifications: notifications.results.map((notification) =>
    mapNotification(notification, type)
  ),
  next: notifications.next?.replace(/.*page=/, "") || undefined,
});

const mapNotification = (notification, type) => ({
  id: notification.slug,
  path: getPath(notification.app, notification.object_id, type),
  title: notification.text,
  time: new Date(notification.timestamp),
  unread: notification.unread,
});

const getPath = (appName, id, type) => {
  const isStandard = type === userTypes.STANDARD;
  if (appName.match(/^cv/)) {
    return isStandard
      ? compile(paths.MY_CVS)({})
      : compile(paths.CV_APPROVAL)({});
  } else if (appName.match(/^jobs/)) {
    return isStandard
      ? compile(paths.JOB_OFFERS)({})
      : compile(paths.MY_OFFERS)({});
  } else if (appName.match(/^account/)) {
    return compile(paths.USER)({});
  } else if (appName.match(/^blog/)) {
    return compile(paths.BLOG_POST)({ id });
  } else {
    return compile(paths.DASHBOARD)({});
  }
};

export const NotificationsContext = createContext(null);

export const NotificationsProvider = (props) => {
  const [notifications, setNotifications] = useState(null);
  const [count, setCount] = useState(null);
  const [next, setNext] = useState();
  const [error, setError] = useState(false);
  const socket = useRef();

  const user = useContext(UserContext);

  useEffect(() => {
    const loadNotifications = async (token, type) => {
      setError(false);
      let values;
      try {
        values = await Promise.all([
          getNotifications(token, type),
          getNotificationsCount(token),
        ]);
      } catch (e) {
        console.log(e);
        setError(true);
        return;
      }
      const [
        { notifications: loadedNotifications, next: loadedNext },
        loadedCount,
      ] = values;
      setCount(loadedCount);
      setNext(loadedNext);
      setNotifications(loadedNotifications);
    };
    if (user.token && user.type) {
      loadNotifications(user.token, user.type);
    }
  }, [user.token, user.type]);

  useEffect(() => {
    if (socket.current) {
      socket.current.close();
      socket.current = null;
    }
    if (user.token && user.type) {
      const url = proxy.wsNotification + "ws";
      try {
        socket.current = new WebSocket(url, user.token);
        // socket.current.onopen = (e) => console.log("onopen", e);
        socket.current.onmessage = (e) => {
          const newNotification = mapNotification(
            JSON.parse(e.data),
            user.type
          );
          setCount((prev) => prev + 1);
          setNotifications((prev) => [newNotification, ...prev]);
        };
        socket.current.onerror = (e) => {
          console.log(e);
          setError(true);
        };
      } catch (e) {
        console.log(e);
      }
    }
  }, [user.token, user.type]);

  const data = {
    notifications,
    count,
    error,
    next,
    socket,
    markAsRead: async () => {
      setCount(0);
      try {
        await markAsRead(user.token);
      } catch (e) {
        setError(true);
        return;
      }
    },
    deleteNotification: async (id) => {
      const notificationToRemove = notifications.find(
        (notification) => notification.id === id
      );
      if (notificationToRemove?.unread && count > 0) {
        setCount((prev) => prev - 1);
      }
      setNotifications((prev) =>
        prev.filter((notification) => notification.id !== id)
      );
      try {
        await deleteNotification(user.token, id);
      } catch (e) {
        setError(true);
        return;
      }
    },
    deleteNotifications: async () => {
      setCount(0);
      setNotifications([]);
      try {
        await deleteNotifications(user.token);
      } catch (e) {
        setError(true);
        return;
      }
    },
    loadMoreNotifications: async () => {
      setNext(null);
      let res;
      try {
        res = await getNotifications(user.token, user.type, next);
      } catch (e) {
        setError(true);
        return;
      }
      setNext(res.next);
      setNotifications((prev) => [...prev, ...res.notifications]);
    },
  };
  return <NotificationsContext.Provider value={data} {...props} />;
};
