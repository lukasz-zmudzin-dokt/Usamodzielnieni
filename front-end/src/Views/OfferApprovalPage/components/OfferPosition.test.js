import React from "react";
import {
  render,
  waitForElement,
  fireEvent,
  wait,
} from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import OfferPosition from "./OfferPosition";
import { AlertContext } from "context";
import proxy from "config/api";

describe("OfferPosition", () => {
  let fetchType;
  let alertC = {
    showAlert: jest.fn(),
  };
  let offer = {
    category: "IT",
    company_address: {
      city: "Wwa",
      postal_code: "21-371",
      street: "Polna",
      street_number: "225",
    },
    company_name: "Firemka",
    description:
      "Szukamy lajkoników w mediach spolecznosciowych. Oferujemy dobra place, karte Multisport...",
    expiration_date: "2020-12-21",
    id: "sadgergerfwefwe",
    offer_name: "Lajkonik",
    type: "Praca",
    voivodeship: "mazowieckie",
  };

  beforeAll(() => {
    global.fetch = jest.fn().mockImplementation(() => {
      return new Promise((resolve, reject) => {
        switch (fetchType) {
          case "approveOk":
            resolve({
              status: 200,
              json: () =>
                Promise.resolve({
                  message: "Ustawiono potwierdzenie oferty pracy",
                }),
            });
            break;
          case "rejectOk":
            resolve({
              status: 200,
              json: () =>
                Promise.resolve({ message: "Offer removed successfully" }),
            });
            break;
          case "odd":
            resolve({
              status: 400,
              json: () =>
                Promise.resolve({
                  message: "Suma podstawy równa się kwadratowi obu ramion",
                }),
            });
            break;
          case "fail":
            resolve({ status: 500 });
            break;
          default:
            reject();
            break;
        }
      });
    });
  });

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("should match snapshot", async () => {
    const { container, getByText } = render(
      <MemoryRouter>
        <OfferPosition offer={offer} />
      </MemoryRouter>
    );
    await waitForElement(() => getByText("mazowieckie"));
    expect(container).toMatchSnapshot();
  });

  it("should approve offer", async () => {
    fetchType = "approveOk";
    const { getByText } = render(
      <AlertContext.Provider value={alertC}>
        <MemoryRouter>
          <OfferPosition offer={offer} />
        </MemoryRouter>
      </AlertContext.Provider>
    );
    await waitForElement(() => getByText("Akceptuj"));
    fireEvent.click(getByText("Akceptuj"));
    await expect(fetch).toHaveBeenCalledWith(
      proxy.job + "admin/confirm/sadgergerfwefwe/",
      {
        body: '{"confirmed":true}',
        headers: {
          Authorization: "token undefined",
          "Content-Type": "application/json",
        },
        method: "POST",
      }
    );
    await wait(() => expect(alertC.showAlert).toHaveBeenCalled());
    expect(alertC.showAlert).toHaveBeenCalledWith(
      "Pomyślnie potwierdzono ofertę",
      "success"
    );
  });

  it("should view error when approving offer fails", async () => {
    fetchType = "fail";
    const { getByText } = render(
      <AlertContext.Provider value={alertC}>
        <MemoryRouter>
          <OfferPosition offer={offer} />
        </MemoryRouter>
      </AlertContext.Provider>
    );
    await waitForElement(() => getByText("Akceptuj"));
    fireEvent.click(getByText("Akceptuj"));

    await wait(() => expect(alertC.showAlert).toHaveBeenCalled());
    expect(alertC.showAlert).toHaveBeenCalledWith(
      "Nie udało się potwierdzić oferty."
    );
  });

  it("should view error when approving offer returns odd api response", async () => {
    fetchType = "odd";
    const { getByText } = render(
      <AlertContext.Provider value={alertC}>
        <MemoryRouter>
          <OfferPosition offer={offer} />
        </MemoryRouter>
      </AlertContext.Provider>
    );
    await waitForElement(() => getByText("Akceptuj"));
    fireEvent.click(getByText("Akceptuj"));
    await wait(() => expect(alertC.showAlert).toHaveBeenCalled());
    expect(alertC.showAlert).toHaveBeenCalledWith(
      "Nie udało się potwierdzić oferty."
    );
  });

  it("should reject offer", async () => {
    fetchType = "rejectOk";
    const { getByText } = render(
      <AlertContext.Provider value={alertC}>
        <MemoryRouter>
          <OfferPosition offer={offer} />
        </MemoryRouter>
      </AlertContext.Provider>
    );
    await waitForElement(() => getByText("Odrzuć"));
    fireEvent.click(getByText("Odrzuć"));
    await expect(fetch).toHaveBeenCalledWith(
      proxy.job + "job-offer/sadgergerfwefwe/",
      {
        headers: {
          Authorization: "token undefined",
          "Content-Type": "application/json",
        },
        method: "DELETE",
      }
    );
    await wait(() => expect(alertC.showAlert).toHaveBeenCalled());
    expect(alertC.showAlert).toHaveBeenCalledWith(
      "Pomyślnie odrzucono ofertę.",
      "success"
    );
  });

  it("should view error when rejecting offer fails", async () => {
    fetchType = "fail";
    const { getByText } = render(
      <AlertContext.Provider value={alertC}>
        <MemoryRouter>
          <OfferPosition offer={offer} />
        </MemoryRouter>
      </AlertContext.Provider>
    );
    await waitForElement(() => getByText("Odrzuć"));
    fireEvent.click(getByText("Odrzuć"));
    await wait(() => expect(alertC.showAlert).toHaveBeenCalled());
    expect(alertC.showAlert).toHaveBeenCalledWith(
      "Nie udało się odrzucić oferty."
    );
  });

  it("should view error when rejecting offer returns odd api response", async () => {
    fetchType = "odd";
    const { getByText } = render(
      <AlertContext.Provider value={alertC}>
        <MemoryRouter>
          <OfferPosition offer={offer} />
        </MemoryRouter>
      </AlertContext.Provider>
    );
    await waitForElement(() => getByText("Odrzuć"));
    fireEvent.click(getByText("Odrzuć"));
    await wait(() => expect(alertC.showAlert).toHaveBeenCalled());
    expect(alertC.showAlert).toHaveBeenCalledWith(
      "Nie udało się odrzucić oferty."
    );
  });
});
