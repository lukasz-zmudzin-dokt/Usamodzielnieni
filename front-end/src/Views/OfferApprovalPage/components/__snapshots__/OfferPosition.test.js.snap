// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OfferPosition should match snapshot 1`] = `
<div>
  <div
    class="p-0 card-body"
  >
    <div
      class="list-group list-group-flush"
    >
      <div
        class="list-group-item"
      >
        <div
          class="row"
        >
          <div
            class="detailsItem col-xl-2 col-md-4"
          >
            <div
              class="detailsItem__label"
            >
              Województwo
            </div>
            <div
              class="detailsItem__value"
            >
              mazowieckie
            </div>
          </div>
          <div
            class="detailsItem col-xl-2 col-md-4"
          >
            <div
              class="detailsItem__label"
            >
              typ
            </div>
            <div
              class="detailsItem__value"
            >
              Praca
            </div>
          </div>
          <div
            class="detailsItem col-xl-2 col-md-4"
          >
            <div
              class="detailsItem__label"
            >
              Firma
            </div>
            <div
              class="detailsItem__value"
            >
              Firemka
            </div>
          </div>
          <div
            class="detailsItem col-xl-2 col-md-4"
          >
            <div
              class="detailsItem__label"
            >
              Adres
            </div>
            <div
              class="detailsItem__value"
            >
              <p>
                Polna
                 
                225
              </p>
              <p>
                21-371
                 
                Wwa
              </p>
            </div>
          </div>
          <div
            class="detailsItem col-xl-2 col-md-4"
          >
            <div
              class="detailsItem__label"
            >
              kategoria
            </div>
            <div
              class="detailsItem__value"
            >
              IT
            </div>
          </div>
          <div
            class="detailsItem col-xl-2 col-md-4"
          >
            <div
              class="detailsItem__label"
            >
              Data wygaśnięcia
            </div>
            <div
              class="detailsItem__value"
            >
              2020-12-21
            </div>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="detailsItem col-xl-3 col-md-2"
          >
            <div
              class="detailsItem__label"
            >
              Wynagrodzenie
            </div>
            <div
              class="detailsItem__value"
            >
              1
               zł - 
              2137
               zł
            </div>
          </div>
        </div>
      </div>
      <div
        class="list-group-item"
      >
        <div
          class="detailsItem col"
        >
          <div
            class="detailsItem__label"
          >
            Opis
          </div>
          <div
            class="detailsItem__value"
          >
            Szukamy lajkoników w mediach spolecznosciowych. Oferujemy dobra place, karte Multisport...
          </div>
        </div>
      </div>
      <div
        class="list-group-item"
      >
        <div
          class="justify-content-center row"
        >
          <button
            class="btn btn-primary"
            type="button"
          >
            Akceptuj
          </button>
          <button
            class="ml-3 btn btn-danger"
            type="button"
          >
            Odrzuć
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
